drivers/gpio/gpio_emul.c

```diff
@@ -0,0 +1,687 @@
+	/* in push-pull, set input values but do not fire interrupts (yet) */
+	gpio_emul_input_set_masked_pend(port, pins & get_input_pins(port), drv_data->output_vals, false);
+	k_mutex_unlock(&drv_data->mu);
+	/* for output-wiring, so the user can take action based on ouput */
+	gpio_fire_callbacks(&drv_data->callbacks, port, pins);
+
+	return 0;
+}
+
+static bool gpio_emul_level_trigger_supported(const struct gpio_emul_config *config,
+					     enum gpio_int_trig trig)
+{
+	switch (trig) {
+	case GPIO_INT_TRIG_LOW:
+		return (config->interrupt_caps & GPIO_EMUL_INT_CAP_LEVEL_LOW) != 0;
+	case GPIO_INT_TRIG_HIGH:
+		return (config->interrupt_caps & GPIO_EMUL_INT_CAP_LEVEL_HIGH) != 0;
+	case GPIO_INT_TRIG_BOTH:
+		return (config->interrupt_caps & GPIO_EMUL_INT_CAP_LEVEL_BOTH) != 0;
+	default:
+		return false;
+	}
+}
+
+static bool gpio_emul_edge_trigger_supported(const struct gpio_emul_config *config,
+					    enum gpio_int_trig trig)
+{
+	switch (trig) {
+	case GPIO_INT_TRIG_LOW:
+		return (config->interrupt_caps & GPIO_EMUL_INT_CAP_LEVEL_LOW) != 0;
+	case GPIO_INT_TRIG_HIGH:
+		return (config->interrupt_caps & GPIO_EMUL_INT_CAP_LEVEL_HIGH) != 0;
+	case GPIO_INT_TRIG_BOTH:
+		return (config->interrupt_caps & GPIO_EMUL_INT_CAP_LEVEL_BOTH) != 0;
+	default:
+		return false;
+	}
+}
+
+static int gpio_emul_pin_interrupt_configure(const struct device *port, gpio_pin_t pin,
+					    enum gpio_int_mode mode,
+					    enum gpio_int_trig trig)
+{
+	int ret;
+	struct gpio_emul_data *drv_data =
+		(struct gpio_emul_data *)port->data;
+	const struct gpio_emul_config *config =
+		(const struct gpio_emul_config *)port->config;
+
+	if ((BIT(pin) & config->common.port_pin_mask) == 0) {
+		return -EINVAL;
+	}
+
+	if (mode != GPIO_INT_MODE_DISABLED) {
+		switch (trig) {
+		case GPIO_INT_TRIG_LOW:
+		case GPIO_INT_TRIG_HIGH:
+		case GPIO_INT_TRIG_BOTH:
+			break;
+		default:
+			return -EINVAL;
+		}
+	}
+
+	if (mode == GPIO_INT_MODE_LEVEL) {
+		if (!gpio_emul_level_trigger_supported(config, trig)) {
+			return -ENOTSUP;
+		}
+	}
+
+	if (mode == GPIO_INT_MODE_EDGE) {
+		if (!gpio_emul_edge_trigger_supported(config, trig)) {
+			return -ENOTSUP;
+		}
+	}
+
+	k_mutex_lock(&drv_data->mu, K_FOREVER);
+
+	switch (mode) {
+	case GPIO_INT_MODE_DISABLED:
+		drv_data->flags[pin] &= ~GPIO_EMUL_INT_BITMASK;
+		drv_data->flags[pin] |= GPIO_INT_DISABLE;
+		break;
+	case GPIO_INT_MODE_LEVEL:
+	case GPIO_INT_MODE_EDGE:
+		drv_data->flags[pin] &= ~GPIO_EMUL_INT_BITMASK;
+		drv_data->flags[pin] |= (mode | trig);
+		break;
+	default:
+		ret = -EINVAL;
+		goto unlock;
+	}
+
+	ret = 0;
+
+unlock:
+	k_mutex_unlock(&drv_data->mu);
+
+	return ret;
+}
+
+static int gpio_emul_manage_callback(const struct device *port,
+				    struct gpio_callback *cb, bool set)
+{
+	struct gpio_emul_data *drv_data =
+		(struct gpio_emul_data *)port->data;
+
+	return gpio_manage_callback(&drv_data->callbacks, cb, set);
+}
+
+static uint32_t gpio_emul_get_pending_int(const struct device *dev)
```

Comment:
s/uint32_t/gpio_port_pins_t/

Reply:
Also in `gpio_emul_flags_get()` and related API (apparently that comment didn't get saved in the review).